Documentation - Extending Wagtail client-side with Stimulus

lb- opened the pull request:
## Introduce a `Extending client-side behaviour` page

- Move the React docs to this new page (these seemed out of place after the direction to have Stimulus under `extending/`).
- Add some introduction content and provide a starting point for adding JavaScript, then using events, then using Stimulus as not all client-side customisations will need to use Stimulus.
- Add a set of initial documentation that let's developers build a basic controller, directions on using (or not using) build tools and more advanced items such as testing, debugging and complex overrides.
- I have done my best to test all examples, the JS ones specifically have been tested on Chrome 114, Safari 16.1, Firefox 113

## Refine the `window.Stimulus` approach

- The previous, undocumented, approach was for the base `Controller` and our `createController` util to be on the application constructor.
- As I started writing docs, this felt confusing and cumbersome, plus it means that we are mutating the Stimulus application class which is not ideal if others want to use this to create additional Stimulus instances.
- We now expose two globals as per feedback;
  - `window.Stimulus` (Stimulus module, as is)
  - `window.stimulusApp` (our Stimulus application instance)
- Rename root element variable to `root` in initStimulus (so we do not conflict with potential action option registration that uses `element` variable names)

## Links

- Builds on #10507 (the initial refactor of naming consistency of client-side vs clientside)
- Closes #10197
- This PR replaces #10510
- An earlier draft of the documentation can be found [here](https://github.com/lb-/wagtail/tree/docs/10197-initial-extending-with-stimulus-docs-v2)

## Future enhancements

Not in scope, as this is already a large PR to review.

- Add a 'more examples' or similar sub-page where we can add other examples (live fields, title fields or the regex filter example).
- Add a better event reference for DOM events we are using, best to flesh this out when we have more content to put there.
- Add JS examples for using `media` on `Panel`s, template components and widgets, so that we can provide a documented better approach for init type scripts. I will probably do that as part of #10311
- A schedule of 'approved' usable Controllers, I thought we best wait until we settle a bit more on naming and edge cases before doing this.

## Checks

_Please check the following:_

-   [X] Do the tests still pass?[^1]
-   [X] Does the code comply with the style guide?
-   [X] For Python changes: Have you added tests to cover the new/fixed behaviour? **N/A**
-   [X] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
-   [X] For new features: Has the documentation been updated accordingly?

SebCorbin commented on docs/extending/extending_client_side.md:
````diff
@@ -0,0 +1,531 @@
+
+Similarly, if some common Stimulus Controller usage would be helpful for the community, such as additional values, targets, DOM events or methods, please raise an issue with the suggestion.
+
+```{warning}
+Usage of `class ... extends` from the globally provided `window.Stimulus.use.Controller` cannot be done outside of code that is also bundled to ES5.
+```
````
This seems to be repeated in the document, is it intentional?

lb- replied:
Thanks, this line was actually incorrect (before we moved to ES6+ targets), so I have removed it.

SebCorbin commented on docs/extending/extending_client_side.md:
```diff
@@ -0,0 +1,531 @@
+
+For more extreme customisations it is possible to retrieve an existing registered Controller from the Stimulus application instance. This can be extended (using ES2015 class inheritance) and this extended controller registered again with the same identifier.
+
+There may also be some side effects of built in controllers being registered, depending on the timing of your JavaScript code event firing.
```
```suggestion
There may also be some side effects of built-in controllers being registered, depending on the timing of your JavaScript code event firing.
```

SebCorbin commented on docs/extending/extending_client_side.md:
```diff
@@ -0,0 +1,531 @@
+
+It is important to note that your custom controller will need to re-implement the existing methods or you will get console errors when these are called.
+
+There may also be some side effects of built in controllers being registered, depending on the timing of your JavaScript code event firing.
```
```suggestion
There may also be some side effects of built-in controllers being registered, depending on the timing of your JavaScript code event firing.
```

Side note: these two last parts seems a lot alike, I wonder if it is useful to keep both.

SebCorbin reviewed (commented):
I wanted to write a RegexController to sync two fields while replacing some content and this is the doc I missed so I made a quick review.

As for the hyphens, consider me not a grammar nazi but I saw [here](https://www.wordreference.com/enfr/out%20of%20the%20box) that these kind of adjectives take hyphens if used before the noun.

Thanks a lot for your work, I understood everything :)

lb- commented:
Thanks for the feedback @SebCorbin - I'll process the items when I get the chance. Good call about the hyphen usage, I do tend to overuse hyphens haha. 

In regards to your use case for syncing fields, I recommend you read the code we are using in our [w-sync controller](https://github.com/wagtail/wagtail/blob/main/client/src/controllers/SyncController.ts) plus the discussion on a proposal for a [SyncFieldPanel](https://github.com/wagtail/wagtail/issues/161#issuecomment-1614211868).

lb- commented:
Thanks @SebCorbin - changes applied, I have also passed this whole page through Grammarly and found a few more glaring mistakes which have been fixed.

Scotchester commented on docs/extending/extending_client_side.md:
```diff
@@ -0,0 +1,527 @@
+
+The simplest way is to add global JavaScript files via hooks, see [](insert_editor_js) and [](insert_global_admin_js).
+
+For JavaScript added when a specific Widget is used you can add an inner `Media` class to ensure that the file is loaded when the widget is used, see [Django form Media](https://docs.djangoproject.com/en/stable/topics/forms/media/#assets-as-a-static-definition).
```
```suggestion
For JavaScript added when a specific Widget is used you can add an inner `Media` class to ensure that the file is loaded when the widget is used, see [Django's docs on their form `Media` class](https://docs.djangoproject.com/en/stable/topics/forms/media/#assets-as-a-static-definition).
```

lb- replied:
Much better thanks.

Scotchester reviewed (dismissed):
This is mostly a driveby – I skimmed the detailed parts and don't have time to run any of the test code – but hopefully the suggestions for improving the intro parts are welcome. The overall organization is spot-on!

lb- commented:
Thanks @Scotchester - super helpful feedback.

lb- commented:
Rebased and updated with some grammar fixes recommended by vale, tried to remove some of the passive voice and used conjunctions.

lb- commented:
Have an approval on this, but will give @thibaudcolas another day or so to review, otherwise will merge in for the 5.1 release freeze.

We can still make tweaks to the docs if needed during the RC period Thibaud but as per core team meeting last week this has a JS change we want to get in before the freeze.

thibaudcolas commented on docs/extending/extending_client_side.md:
````diff
@@ -0,0 +1,528 @@
+You don't need to have Node.js tooling running for your custom Wagtail installation for many customisations built on these libraries, but in some cases, such as building packages, it may make more complex development easier.
+
+```{note}
+it's recommended that you avoid using jQuery as this will be removed in a future version of Wagtail.
````
If we have this kind of long-term note I feel like we might as well make it a clear-cut "don’t".

```suggestion
Avoid using jQuery and any undocumented jQuery plugins, as they will be removed in a future version of Wagtail.
```

thibaudcolas commented on docs/extending/extending_client_side.md:
````diff
@@ -0,0 +1,528 @@
+For example, if you want to attach some logic to a field value change in Wagtail you can add an event listener, check if it's the correct element and change what you need.
+
+```javascript
+// myapp/static/js/events.js
+document.addEventListener('change', function (event) {
+    if (event.target) {
````
This feels dated to me to add global event listeners to interact with a single field. Is this so people can implement those changes without worrying about the execution order of the JS initialisation?

lb- replied:
The idea is that JS integrations do not need to be complex and event listeners are an elegant, simple, maintainable way to get many kinds of customisations working. 

Additionally, it paves the way for Stimulus usage which is built on Events as first class citizens.

I'd also push back a bit on event listeners being dated, I think it's a very useful way to add behaviour that the browser is built around.

laymonage replied:
I think Thibaud's point is about global event listeners, not event listeners in general. Adding global event listeners to specific elements should be fine, but adding event listeners to the document element means that the function is executed even when the event comes from an element that is not of the function's concerns (which I think is what Thibaud meant by "wasteful" in another comment).

There are cases where this is useful, e.g. capturing a "click away", but in most cases I don't think it's a good idea to advise this.

lb- replied:
Thanks both, I will just remove this example and keep this section short and generic.

The underlying goal is to remind developers that they may not need to reach for a complex solution when something simple may get the job done.

thibaudcolas commented on docs/extending/extending_client_side.md:
````diff
@@ -0,0 +1,528 @@
+
+```javascript
+// myapp/static/js/events.js
+document.addEventListener('click', function (event) {
+    if (!event.target) return;
+
+    isCommentsToggle = event.target.matches(
+        '[data-side-panel-toggle="comments"]',
+    );
````
I’m finding this problematic for the above reason (executing JS on every click seems wasteful / dated), and also because we’re now document those `data-` attributes as things you can rely on in your customisation. I’d prefer we stuck to documented parts of Wagtail, or at least examples that are less tied to our implementation details.

If we keep this,

```suggestion
document.addEventListener('click', function (event) {
    if (!event.target) return;

    const isCommentsToggle = event.target.matches(
        '[data-side-panel-toggle="comments"]',
    );
```

lb- replied:
I'll try to think of a better example here. I agree we probably don't want to highlight this data attribute specifically. 

However, as above, I'd push back on this being dated. Remember the target audience is people trying to solve things they cannot solve with html / Python alone.

lb- replied:
As per comment above, will remove this example, keep it generic and just focus on the goal of reminding devs that Events exist.

thibaudcolas commented on docs/extending/extending_client_side.md:
```diff
@@ -0,0 +1,528 @@
+
+### Wagtail's custom DOM events
+
+-   Wagtail supports some custom behaviour via listening or dispatching custom DOM events, usually with the prefix `wagtail:` or `w-` for specific Stimulus controllers.
```
Here I think my preference would be to document the specific events we want people to reuse, rather than setting a pattern of allowing reliance on any and all DOM events with those prefixes. It feels like too much busywork for us to treat all prefixed DOM events as needing to remain API-compatible. If an event is important enough for third-party customisations, then I think having it explicitly named and described in the docs is a good baseline.

lb- replied:
Great, I'll remove this.

thibaudcolas commented on docs/extending/extending_client_side.md:
```diff
@@ -0,0 +1,528 @@
+
+## Extending with Stimulus
+
+Wagtail uses [Stimulus](https://stimulus.hotwired.dev/) as a way to provide lightweight client-side interactivity where React isn't required.
```
```suggestion
Wagtail uses [Stimulus](https://stimulus.hotwired.dev/) as a way to provide lightweight client-side interactivity.
```

thibaudcolas commented on docs/extending/extending_client_side.md:
```diff
@@ -0,0 +1,528 @@
+## Extending with Stimulus
+
+Wagtail uses [Stimulus](https://stimulus.hotwired.dev/) as a way to provide lightweight client-side interactivity where React isn't required.
+You can use Stimulus to easily build custom JavaScript widgets within the admin interface.
```
```suggestion
You can use Stimulus to build custom JavaScript widgets within the admin interface.
```

thibaudcolas commented on docs/extending/extending_client_side.md:
```diff
@@ -0,0 +1,528 @@
+
+Wagtail uses [Stimulus](https://stimulus.hotwired.dev/) as a way to provide lightweight client-side interactivity where React isn't required.
+You can use Stimulus to easily build custom JavaScript widgets within the admin interface.
+The key benefit of using Stimulus is that your code can avoid the need for manual initialisation when widgets appear dynamically in the browser. Such as within modals, `InlinePanels` or `StreamField` panels.
```
```suggestion
The key benefit of using Stimulus is that your code can avoid the need for manual initialisation when widgets appear dynamically in the browser. Such as within modals, `InlinePanel`, or `StreamField` panels.
```

lb- replied:
Thanks

thibaudcolas commented on docs/extending/extending_client_side.md:
```diff
@@ -0,0 +1,528 @@
+You can use Stimulus to easily build custom JavaScript widgets within the admin interface.
+The key benefit of using Stimulus is that your code can avoid the need for manual initialisation when widgets appear dynamically in the browser. Such as within modals, `InlinePanels` or `StreamField` panels.
+
+### Understanding the basics of Stimulus
```
I don’t get why we have all of this in our docs. Wouldn’t people be better off by just looking at the Stimulus docs?

lb- replied:
Yeah, I struggled with this. I found myself referencing Stimulus concepts later and it felt like I should have some basics on this page.

My reasoning is the audience could get away with getting something custom working and understood without reading an external page. 

Happy to remove this completely and just direct people to the Stimulus docs.

thibaudcolas commented on docs/extending/extending_client_side.md:
```diff
@@ -0,0 +1,528 @@
+You need to first create a custom [Stimulus Controller](https://stimulus.hotwired.dev/reference/controllers). This can be done in two main ways, depending on your set-up:
+
+1. Create a **custom Class** that extends the base `window.Stimulus.use.Controller` using class inheritance. If you are using a build tool you can extend `import { Controller } from '@hotwired/stimulus';`.
+2. If you are unable to work with classes, build one with `window.Stimulus.use.createController` which accepts **an object using the [method definitions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Method_definitions) approach**, with a special `STATIC` property for static properties.
```
I’d rather we remove this to be honest. ES6 classes are a good 8 years old and weren’t particularly complex to understand even back then. The more modern `static` syntax isn’t that much harder to understand and has been around for a while too.

lb- replied:
No problems, I think there may be an edge case here and there that will warrant the usage of this object until. However, it's probably just confusing to have this documented here.

lb- replied:
Ignoring edge cases, if devs get stuck we can re-add this later or provide some example code for a work around.

thibaudcolas commented on docs/extending/extending_client_side.md:
```diff
@@ -0,0 +1,528 @@
+
+### Adding a custom Stimulus controller
+
+Wagtail exposes a single client-side global (`window.Stimulus`), which is an instance of the core admin Stimulus application.
```
How about we instead expose two globals?

```js
import * as Stimulus from "@hotwired/stimulus"

window.Stimulus = Stimulus;
window.stimulusApp = Stimulus.Application.start();
```

This goes against the examples in the Stimulus docs, but makes it much clearer which is a package and which is the instance, while avoiding the need for a custom `.use` which I think could easily be confused with an official API.

In the future this will also make it easier for us to support `import { Controller } from 'stimulus'` in ES modules.

lb- replied:
I'm confused by this. 

All the documentation I've read uses `window.Stimus` as the application instance and never `window.stimulusApp`. 

https://stimulus.hotwired.dev/handbook/installing#using-other-build-systems

Originally I had lowercase `window.application` for the instance (early PRs) and you advised this should follow the docs and be `window.Stimulus`.

I'm not sure what to do with exposing a base a `Controller`, I've thought about this a fair bit and tried the docs with a lot of variants of this and `Stimulus.use.Controller` felt the simplest. 

I agree we would love to see this as an es6 import but that's a while away and won't work with widjet media JS scripts. 

If you're pretty set on not having anything else on the application that may be a 'box' for us to put custom stuff in, maybe some other ideas;

- expose it as a standalone global
- expose it on the bare stimulus application instance

I'm just trying to be aware of the spectrum of users of this stuff but understand the goal of not confusing our Stimulus usage with the core docs.

lb- replied:
Have opted for your suggestion Thibaud, scrapping the `createController` util completely so we can just expose the module and then the application instance.

I would like `window.app` or `window.application` but will update the code to be `window.stimulusApp` as recommended. My reasoning is that `app` is shorter but I understand this may confuse.

thibaudcolas reviewed (changes requested):
I’m not convinced with the direction here :/ I have a feeling this is setting us up for a lot of  maintenance burden, and the demand from people making customisations hasn’t been present enough to justify this.

thibaudcolas commented on docs/extending/extending_client_side.md:
```diff
@@ -0,0 +1,528 @@
+// myapp/static/js/example.js
+
+// 1. With classes
+(() => {
```
IMO showing this with a IIFE / proper closure should be out of scope for our docs. All those examples would be clearer without the extra level of indentation and JS idioms.

lb- replied:
Agreed, removing, may pollute globals if done without the function wrapper but not a problem for simplicity.

thibaudcolas commented on docs/extending/extending_client_side.md:
```diff
@@ -0,0 +1,528 @@
+        setupOutput() {
+            if (this.output) return;
+            const template = document.createElement('template');
+            template.innerHTML = `<output name='word-count' for='${this.element.id}' style='float: right;'></output>`;
```
I think we should avoid inline styles in example code in the spirit of CSP compatibility.

thibaudcolas commented on docs/extending/extending_client_side.md:
````diff
@@ -0,0 +1,528 @@
+    #...
+```
+
+This next code snippet shows a slightly more advanced version of the `insert_editor_js` hook usage which is set up to easily append additional scripts for future controllers.
````
```suggestion
This next code snippet shows a more advanced version of the `insert_editor_js` hook usage which is set up to append additional scripts for future controllers.
```

thibaudcolas commented on docs/extending/extending_client_side.md:
````diff
@@ -0,0 +1,528 @@
+    )
+```
+
+You should be able to see that on your Blog Page, all introduction fields will now have a small `output` element showing the count and max words being used.
````
Wouldn’t there only be one field?

```suggestion
You should be able to see that on your Blog Page, the introduction field will now have a small `output` element showing the count and max words being used.
```

Or this needs to say "blog pages".

thibaudcolas commented on docs/extending/extending_client_side.md:
```diff
@@ -0,0 +1,528 @@
+
+You should be able to see that on your Blog Page, all introduction fields will now have a small `output` element showing the count and max words being used.
+
+#### A word count controller example with a build system
```
This strikes me as way too much information compared to what we definitely need to convey here:

- Use Wagtail’s Stimulus package with our global variable(s) if possible, or your own copy of `@hotwired/stimulus` if needed.
- Make sure to use a build target higher than ES5

thibaudcolas commented on docs/extending/extending_client_side.md:
````diff
@@ -0,0 +1,528 @@
+### Using an existing Stimulus controller in HTML
+
+```{warning}
+While many Stimulus controllers are in use, this doesn't imply stable and documented usage is supported for all.
````
This – and below "Any built-in admin Stimulus Controller can be used via the Stimulus data attributes." –  is the kinds of statements I don’t see working too well for us. I’d rather we specifically list which controllers should be available for reuse, document those, and keep their API stable.

lb- replied:
I will just remove this whole section, I am not sure it's worth pushing this for the initial documentation. I agree that we need to be careful here with what we document and based on other PRs it's more likely we will expose (and document) a Django widget, Panel or even template tag before we document too many 'low level' implementation items like this.

thibaudcolas commented on docs/extending/extending_client_side.md:
````diff
@@ -0,0 +1,528 @@
+While many Stimulus controllers are in use, this doesn't imply stable and documented usage is supported for all.
+```
+
+The Stimulus approach aligns with this philosophy by relying on data attributes that can be often set when declaring Django Widgets and Wagtail Panels in Python or the HTML templates.
````
I’m not clear which philosophy we’re referring to here?

lb- replied:
Will remove this, along with the section, but agree that the term philosophy is not really suitable.

thibaudcolas commented on docs/extending/extending_client_side.md:
````diff
@@ -0,0 +1,528 @@
+```
+
+The Stimulus approach aligns with this philosophy by relying on data attributes that can be often set when declaring Django Widgets and Wagtail Panels in Python or the HTML templates.
+For commonly used Controllers there may already be a Django widget or `FieldPanel` approach that supports your requirements already.
````
The TitleCase "Controller" is confusing to me, should this not be lowercase?

lb- replied:
Agreed, have updated them all to simply 'controller'.

thibaudcolas commented on docs/extending/extending_client_side.md:
```diff
@@ -0,0 +1,528 @@
+-   `<button ... ><strong data-w-progress-target="label">{% trans 'Create' %}</strong></button>` - any element can be the button label (not just `em`)
+-   `<button ... data-action="w-progress#activate focus->w-progress#activate" ...>` - any event can be used to trigger the in-progress behaviour
+-   `<button ... data-action="w-progress#activate:once" ...>` - only trigger the progress behaviour once
+-   `<button ... data-action="readystatechange@document->w-progress#activate:once" data-w-progress-duration-value="5000" disabled ...>` - disabled on load (once JS starts) and becomes enabled after 5s duration
```
It’s not clear to me in what scenarios this is useful?

lb- replied:
As above, will remove this whole section, needs more intentional guidance and we may never want/need to document this low level kind of usage.

thibaudcolas commented on docs/extending/extending_client_side.md:
```diff
@@ -0,0 +1,528 @@
+
+### Testing your Controllers
+
+Wagtail uses an out-of-the-box Stimulus application instance, you should be able to write your tests with your package.json installation of Stimulus without issue.
```
I don’t understand what "out-of-the-box" means here?

lb- replied:
Removing this, devs can work this out themselves, but agree this line was a bit confusing.

lb- commented:
> 

In regards to this, I'm not sure what to do sorry, happy to bail on this PR if we didn't want to document this stuff. I would love some direction as to our goals for what we do want or do not want to document. 

If this isn't really ready for 5.1, all good, it's my first go at this. But would really appreciate some understanding of where to go with this. 

For context on demand, that's a tricky one to quantify, I've probably pointed a handful of people on Slack to this PR (and the first one) since it was raised.

lb- commented:
Note to self. May want to cross reference this page on https://docs.wagtail.org/en/latest/advanced_topics/customisation/streamfield_blocks.html#custom-block-types

lb- commented on docs/extending/extending_client_side.md:
```diff
@@ -0,0 +1,528 @@
+-   `<button ... data-action="w-progress#activate:once" ...>` - only trigger the progress behaviour once
+-   `<button ... data-action="readystatechange@document->w-progress#activate:once" data-w-progress-duration-value="5000" disabled ...>` - disabled on load (once JS starts) and becomes enabled after 5s duration
+
+### Debugging & error handling
```
Fair enough, will remove for now, if questions come up we can append a 'more links' section here in the future. Otherwise, the Stimulus docs has all of this.

lb- commented on docs/extending/extending_client_side.md:
```diff
@@ -0,0 +1,528 @@
+
+For additional guidance on building Stimulus controllers, you can view the [](ui_guidelines_stimulus) aimed at Wagtail contributors.
+
+#### Registering a custom Stimulus application instance
```
agreed, removing.

lb- commented on docs/extending/extending_client_side.md:
````diff
@@ -0,0 +1,528 @@
+})(window.Stimulus);
+```
+
+#### Extending an existing Stimulus controller
````
Agreed, removing.

lb- commented on docs/extending/extending_client_side.md:
````diff
@@ -0,0 +1,528 @@
+})();
+```
+
+#### Completely overriding existing admin behaviour of Stimulus controllers
````
Agreed, removing section.

lb- commented:
@thibaudcolas @Scotchester @laymonage 

Ready for another round of reviews, hopefully I have understood the feedback Thibaud. 

This is now much more cut down, doesn't try to document anything that's already in the Stimulus docs and keeps the focus super tight on just enough to get a custom controller registered.

If we do decide to have more officially supported events and controllers from the admin we can add these in the future.

I've also added a few more cross references to our Telepath JS usage on this new page.

lb- commented:
Rebased and also changed one small thing on the `initStimulus` function. Renamed the root element variable to `root` instead of `element`.

This is so we do not conflict with potential action option registration that uses `element` variable names, I started doing this for the Panel/RevealController work and found that the naming conflict is a bit annoying so best to fix in this PR in prep for that work.

jsma commented on docs/extending/extending_client_side.md:
```diff
@@ -0,0 +1,277 @@
+
+Wagtail exposes two client-side globals for using Stimulus.
+
+1. `window.stimulusApp` the an instance of the core admin Stimulus application.
```
"~~the~~ an"

lb- replied:
Thanks, fixed this and a few other grammar typos found by Grammarly.

jsma commented on docs/extending/extending_client_side.md:
```diff
@@ -0,0 +1,277 @@
+1. `window.stimulusApp` the an instance of the core admin Stimulus application.
+2. `window.Stimulus` Stimulus module as exported from `@hotwired/stimulus`.
+
+You need to first create a custom [Stimulus controller](https://stimulus.hotwired.dev/reference/controllers) which will the base `window.Stimulus.Controller` using class inheritance. If you are using a build tool you can extend `import { Controller } from '@hotwired/stimulus';`.
```
"which will the base `window.Stimulus.Controller` using class inheritance." => "which should extend `window.Stimulus.Controller` using class inheritance."

lb- replied:
Thanks, went with a slightly different wording of this whole paragraph, hopefully reads better now.

thibaudcolas commented:
Just noting we have lots of reviewers on this currently – I’ve committed to reviewing this PR in a core team meeting, and this will happen _next week_. I think there’s a fair chance I’d approve with minimal to no changes then, so if anyone wants to review nonetheless please go for it.

thibaudcolas reviewed (approved):
@lb- this is all looking excellent. I think it’ll be a solid foundation for package-level and site-level customisations!

Before merging, I thought I’d raise the choice of global variable names one last time, as it’s something that would be annoying for us to have to change in the future. I’m happy with the variables as-is but have provided alternatives in case you want to consider options, with the specific goal of respecting the Stimulus docs’ usage of `Stimulus`.

lb- commented on client/src/entrypoints/admin/core.js:
```diff
@@ -1,11 +1,13 @@
 /** initialise Wagtail Stimulus application with core controller definitions */
-window.Stimulus = initStimulus({ definitions: coreControllerDefinitions });
+window.stimulusApp = initStimulus({ definitions: coreControllerDefinitions });
+window.Stimulus = Stimulus;
```
I'm on board for

- `window.wagtail.app` for the Stimulus application instance. 
- `windoe.StimulusModule` for the module exposed export.

This avoids naming confusion or collisions with `window.Stimulus`/`window.stimulus`. Which means the devs will encounter this as the 'wrong' way much easier as it will be undefined. 

Let me know if you agree and I'm happy to make the change.

lb- replied:
After a bit of thinking, I feel that this is a much better approach.

We already have `window.wagtail.components` for our exposed React components, I think that this gives us a nicer way to expose globals long term if needed.

The `StimulusModule` on it's own is a good approach, maybe we can remove this one day and provide a nicer module import approach (if we can make this work easily with Django static JS approaches).

I will update this PR accordingly.

I have a backup of the current approach if needed though.

lb- replied:
Ok. This is done. I've also moved the Wagtail object stuff into the one file, added unit tests for how we set up.this object and added JSDoc. 

Happy to move this all to wagtailadmin.js instead but maybe that's best done as a seperate issue when a few more things are migrated away from core/Wagtail admin.js

lb- replied:
@thibaudcolas  when you have a min, could you look overt this PR again and let me know if we are good to go with these last set of naming changes applied.

lb- commented:
Screenshot validation of new globals being used.

<img width="1964" alt="Screenshot 2023-09-28 at 5 56 34 pm" src="https://github.com/wagtail/wagtail/assets/1396140/f5242735-480e-408e-93e0-b9042a649273">


<img width="1969" alt="Screenshot 2023-09-28 at 5 54 46 pm" src="https://github.com/wagtail/wagtail/assets/1396140/29bbf593-2a2c-4889-9bd0-a02e7474470f">

Reminder `window.wagtail` will also be the document's element with id `wagtail` - this is part of the HTML spec and is working as expected.
https://html.spec.whatwg.org/#named-access-on-the-window-object

thibaudcolas reviewed (approved):
Thank you @lb-!

lb- commented:
Thank you for the thoughtful reviews @thibaudcolas  and others, I hope this really helps devs build cool things more simply in the future.
